neutron_fwaas/services/firewall/fwaas_plugin_v2.py

```diff
@@ -162,12 +163,18 @@ def _validate_ports_for_firewall_group(self, context, tenant_id,
         # TODO(sridar): elevated context and do we want to use public ?
         for port_id in fwg_ports:
             port = self._core_plugin.get_port(context, port_id)
-
-            if port['tenant_id'] != tenant_id:
+            device_owner = port.get('device_owner', '')
+            if port['device_owner'] == nl_constants.DEVICE_OWNER_ROUTER_GW:
+                # The gateway interface has no tenant, test if the gateway interface's router
+                # is owned by the tenant requestor
+                router = self.l3_plugin.get_router(context, port['device_id'])
+                if router['tenant_id'] != tenant_id:
+                    raise f_exc.FirewallGroupPortInvalidProject(
+                        port_id=port_id, project_id=port['tenant_id'])
```

Comment:
How about rewriting this that it just overrides the ports tenant_id?
```
port_tenant_id = port['tenant_id']
if port['device_owner'] == nl_constants.DEVICE_OWNER_ROUTER_GW:
     router = self.l3_plugin.get_router(context, port['device_id'])
    if router:
        port_tenant_id = router['tenant_id']

# original code starts here
if port_tenant_id != tenant_id:
    raise ....

Reply:
Yes, that works. You did an additional check if the router exists, which it should since there is a gw interface. However I believe it can happen that a race condition deleted the router in the meantime. So I'll just throw an exception if the router is gone.

Reply:
I implemented it that [way](https://github.com/sapcc/neutron-fwaas/compare/eddc58611247c2a118cdfcf6e1dcf39e5619d870..0d3f61cc6aa943d5eb6684bdc46f53bd3fef47cd) but I am not sure if I like it more.

Reply:
Both versions work for me. I think the latter represents a bit more how I personally think through problems like this, but I wouldn't necessarily call it better. Therefore feel free to choose what fits your thinking style :)